src/twisted/internet/defer.py

```diff
@@ -295,60 +362,80 @@ def __init__(self, canceller=None):
         @type canceller: a 1-argument callable which takes a L{Deferred}. The
             return result is ignored.
         """
-        self.callbacks = []
+        self.callbacks = []  # type: List[_Callbacks]
         self._canceller = canceller
         if self.debug:
             self._debugInfo = DebugInfo()
             self._debugInfo.creator = traceback.format_stack()[:-1]
 
     def addCallbacks(
-        self,
-        callback,
-        errback=None,
-        callbackArgs=None,
-        callbackKeywords=None,
-        errbackArgs=None,
-        errbackKeywords=None,
-    ):
+        self: _DeferredT,
+        callback: DeferredCallback,
+        errback: DeferredErrback = passthru,
+        callbackArgs: _CallbackOrderedArguments = (),
+        callbackKeywords: _CallbackKeywordArguments = _NONE_KWARGS,
+        errbackArgs: _CallbackOrderedArguments = (),
+        errbackKeywords: _CallbackKeywordArguments = _NONE_KWARGS,
+    ) -> _DeferredT:
```

Comment:
In addition to self-as-typevar having some issues with mypy's internals, there's also the issue here that this signature is conceptually wrong too; `_DeferredT.addCallback(x->y)`  does *not* return `_DeferredT` in the future generic version, it returns `Deferred[Y]`.  I think if we want to go with simpler stubs for the first iteration, it would be better to simply type this as `Deferred`.  (I think the idea here is to capture `DeferredList`'s subclass behavior, but `DeferredList` being a subclass ought to be irrelevant; the public attributes `fireOnOneCallback` etc. are exposed by accident from the constructor, ought to be private, and it has no differing public behaviors beyond those.

Reply:
I'm not sure why you are objecting to `self`-as-typevar being bad for mypy, since the mypy docs suggest it: https://mypy.readthedocs.io/en/stable/generics.html#generic-methods-and-generic-self

But yes, ok, the signature is wrong.  I'll fix that.

Reply:
Mostly this annotation:

> This feature is experimental. Checking code with type annotations for self arguments is still not fully implemented. Mypy may disallow valid code or allow unsafe code.

Plus some personal experiences with bugs when attempting to do this.  For all I know all the bugs *I* hit are fixed by now but the presence of that warning suggests to me that it's not fully cooked yet.

(I *want* this feature to be done; our work codebase has a bajillion classmethod constructors or generative modifiers which have slightly-inaccurate types because we've been burned by this and we don't want to start depending on it until they at least *claim* it works right)

Reply:
Somehow that box was rendered invisibly to me until now.  :-)

OK… well, that's a real bummer.  OK.  Fine.  Yeah.  OK.  I'll stop doing that for now.